.gitlab/docker/Dockerfile.spack_build

```diff
@@ -0,0 +1,20 @@
+ARG BASE_IMAGE
+FROM $BASE_IMAGE
+
+ARG SOURCE_DIR
+ARG BUILD_DIR
+
+ARG CMAKE_COMMON_FLAGS
+ARG CMAKE_FLAGS
+
+COPY . ${SOURCE_DIR}
+
+# Configure & Build
+RUN spack build-env $spack_spec -- bash -c "cmake -B${BUILD_DIR} ${SOURCE_DIR} \
+    $CMAKE_COMMON_FLAGS $CMAKE_FLAGS && cmake --build ${BUILD_DIR} --target all tests examples"
+
+# Run compile only tests and tests.unit.build
+RUN spack build-env $spack_spec -- bash -c "ctest -L COMPILE_ONLY --test-dir ${BUILD_DIR} \
+    --verbose -j$(nproc) --timeout 30 --output-on-failure --no-compress-output -R tests \
+    && ctest --test-dir ${BUILD_DIR} --verbose -j$(nproc) --timeout 15 --output-on-failure \
+    --no-compress-output -R tests.unit.build"
```

Comment:
Let's go with this for now, but see https://github.com/pika-org/pika/issues/816 for a potential follow-up. I think `nproc` may be overkill here, but it's not a blocker as long as we don't kill the build nodes by using too much memory...